lib/Transforms/Instrumentation/ShadowMemSeaDsa.cc

```diff
@@ -1353,15 +1355,60 @@ void ShadowDsaImpl::solveUses(Function &F) {
                                        << " use(s) solved.\n");
 }
 
-} // namespace
+/// \brief Returns id of a field pointed to by the given cell \c
+unsigned ShadowDsaImpl::getFieldId(const dsa::Cell &c) {
+  assert(c.getNode());
+  unsigned id = m_maxId; // in case allocation of new id for the node is needed
+
+  const sea_dsa::Node *n = c.getNode();
+  unsigned offset = c.getOffset();
+  auto it = m_nodeIds.find(n);
+  if (it != m_nodeIds.end()) {
+    id = it->second + offset;
+  }
+  else{
+    m_nodeIds[n] = id;
+    if (n->size() == 0) {
+      // XXX: what does it mean for a node to have 0 size?
+      assert(offset == 0);
+      ++m_maxId;
+    } else {
+      // -- allocate enough ids for the node and all its fields
+      m_maxId += n->size();
+      id = id + offset;
+    }
+  }
+  auto it2 = m_shadowsReverse.find(id);
+  if(it2 == m_shadowsReverse.end()){ // add to the reverse map if needed
+    m_shadowsReverse.insert(std::make_pair(id, std::make_pair(c.getNode(), c.getRawOffset())));
```

Comment:
You should be able to use `{c.getNode(), c.getRawOffset()}` instead of `std::make_pair`